### PR TITLE
feat: new error types for destinations(marketo, marketo_static_list, marketo_bulk_upload)

### PR DESCRIPTION
**Pull request by mihir-4116**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js [hunk 1]
```diff
@@ -58,62 +63,65 @@ const getFailedJobStatus = async event => {
...
           integration: "Marketo_bulk_upload",
           status: 500,
           state: "Retryable"
         });
-        throw new CustomError(
+        throw new ThrottledError(
           resp.response.response.statusText ||
             "Error during fetching failure job status",
-          500
+          500,
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js:96]: Please check arguments (for throttled no status should be passed)

File: src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js [hunk 2]
```diff
@@ -162,62 +170,65 @@ const getWarningJobStatus = async event => {
...
           integration: "Marketo_bulk_upload",
           status: 500,
           state: "Retryable"
         });
-        throw new CustomError(
+        throw new ThrottledError(
           resp.response.response.statusText ||
             "Error during fetching warning job status",
-          500
+          500,
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js:203]: Remove the status

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 3]
```diff
@@ -46,9 +55,13 @@ const fetchFieldSchema = async config => {
...
   } else if (fieldSchemaMapping.response.error) {
-    throw new CustomError(`${fieldSchemaMapping.response.error}`, 400);
+    throw new NetworkInstrumentationError(
+      `${fieldSchemaMapping.response.error}`
+    );
   } else {
-    throw new CustomError("Failed to fetch Marketo Field Schema", 400);
+    throw new NetworkInstrumentationError(
+      "Failed to fetch Marketo Field Schema"
+    );
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/fileUpload.js:64]: For now lets make these network errors as the logic needs to be refined

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 4]
```diff
@@ -59,14 +72,14 @@ const getHeaderFields = (config, fieldSchemaNames, jobIds) => {
   columnFieldsMapping.forEach(colField => {
     if (fieldSchemaNames) {
       if (fieldSchemaNames && !fieldSchemaNames.includes(colField.to)) {
-        throw new CustomError(
+        throw new AbortedError(
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/fileUpload.js:75]: Config error

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 5]
```diff
@@ -59,14 +72,14 @@ const getHeaderFields = (config, fieldSchemaNames, jobIds) => {
...
           `The field ${colField.to} is not present in Marketo Field Schema. Aborting`,
           400,
           { successfulJobs: jobIds, unsuccessfulJobs: [] }
         );
       }
     } else {
-      throw new CustomError(`Marketo Field Schema is Empty. Aborting. `, 400, {
+      throw new AbortedError("Marketo Field Schema is Empty. Aborting", 400, {
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/fileUpload.js:82]: This is a marketo ecosystem setup problem (out side of rudderstack) shall we attribute as config @saikumarrs ?

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 6]
```diff
@@ -325,7 +338,7 @@ const getImportID = async (input, config, fieldSchemaNames, accessToken) => {
               status: 500,
               state: "Retryable"
             });
-            throw new CustomError(
+            throw new ThrottledError(
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/fileUpload.js:341]: Remove status

File: src/v0/destinations/marketo_bulk_upload/poll.js [hunk 7]
```diff
@@ -77,9 +82,10 @@ const getPollStatus = async event => {
           status: 500,
           state: "Retryable"
         });
-        throw new CustomError(
+        throw new ThrottledError(
           pollStatus.response.data.errors[0].message || "Could not poll status",
-          500
+          500,
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/poll.js:87]: Remove status

File: src/v0/destinations/marketo_bulk_upload/util.js [hunk 8]
```diff
@@ -38,44 +43,52 @@ const getAccessToken = async config => {
...
     }
-    throw new CustomError("Could not retrieve authorisation token", 400);
+    throw new NetworkInstrumentationError(
+      "Could not retrieve authorization token"
+    );
   }
-  throw new CustomError("Could not retrieve authorisation token", 400);
+  throw new NetworkInstrumentationError(
+    "Could not retrieve authorization token"
+  );
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/util.js:91]: Auth failures here Auth errors should be here @saikumarrs can you confirm ?

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js (hunk 6 above)
Comment (saikumarrs) [inline src/v0/destinations/marketo_bulk_upload/fileUpload.js:342]: Same as above.

File: src/v0/destinations/marketo_bulk_upload/util.js [hunk 9]
```diff
@@ -38,44 +43,52 @@ const getAccessToken = async config => {
     ) {
       return resp.response.data.access_token;
     }
-    throw new CustomError("Could not retrieve authorisation token", 400);
+    throw new NetworkInstrumentationError(
```
Comment (saikumarrs) [inline src/v0/destinations/marketo_bulk_upload/util.js:47]: Not an instrumentation error.

File: src/v0/destinations/marketo_bulk_upload/util.js (hunk 8 above)
Comment (saikumarrs) [inline src/v0/destinations/marketo_bulk_upload/util.js]: Not instrumentation errors.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js [hunk 10]
```diff
@@ -252,7 +263,7 @@ const responseHandler = async (event, type) => {
   if (metadata && metadata.csvHeader) {
     headerArr = metadata.csvHeader.split(",");
   } else {
-    throw new CustomError("No csvHeader in metadata", 400);
+    throw new InstrumentationError("No csvHeader in metadata");
```
[the remark this answers is not in the document]
Comment (utsabc) [reply src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js:264]: Yes missed it yes platform error

File: src/v0/destinations/marketo_bulk_upload/util.js (hunk 8 above)
Comment (saikumarrs) [reply src/v0/destinations/marketo_bulk_upload/util.js]: Yes, agreed.

File: src/v0/destinations/marketo_bulk_upload/util.js [hunk 11]
```diff
@@ -82,13 +91,9 @@ const getAccessToken = async config => {
...
     }
-    throw new NetworkInstrumentationError(
-      "Could not retrieve authorization token"
-    );
+    throw new UnauthorizedError("Could not retrieve authorization token");
   }
-  throw new NetworkInstrumentationError(
-    "Could not retrieve authorization token"
-  );
+  throw new UnauthorizedError("Could not retrieve authorization token");
```
Comment (saikumarrs) [inline src/v0/destinations/marketo_bulk_upload/util.js:96]: These are not unauthorized errors. You can treat them as general network errors.

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 12]
```diff
@@ -55,13 +54,19 @@ const fetchFieldSchema = async config => {
...
+      },
+      fieldSchemaMapping
     );
   } else {
-    throw new NetworkInstrumentationError(
-      "Failed to fetch Marketo Field Schema"
-    );
+    throw new NetworkError("Failed to fetch Marketo Field Schema", 400, {
+      [tags.TAG_NAMES.ERROR_TYPE]: getDynamicErrorType(400)
+    });
```
Comment (saikumarrs) [inline src/v0/destinations/marketo_bulk_upload/fileUpload.js:69]: Make it an aborted error. The status is already 400.

**Review by utsabc — APPROVED** [review]

File: src/v0/destinations/marketo/transform.js [hunk 13]
```diff
@@ -518,8 +518,7 @@ const processRouterDest = async inputs => {
     const errResp = {
       status: 400,
       message: "Authorisation failed",
-      responseTransformFailure: true,
-      statTags: {}
```
Comment (saikumarrs) [inline src/v0/destinations/marketo/transform.js]: Retain this.

Comment (mihir-4116) [reply src/v0/destinations/marketo/transform.js]: done

File: src/v0/destinations/marketo_static_list/transform.js [hunk 14]
```diff
@@ -150,9 +132,7 @@ const processRouterDest = async inputs => {
         message: "Authorisation failed",
         responseTransformFailure: true,
         statTags: {
-          destType: DESTINATION,
-          stage: TRANSFORMER_METRIC.TRANSFORMER_STAGE.TRANSFORM,
-          scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.AUTHENTICATION.SCOPE
+          destType: DESTINATION
```
Comment (saikumarrs) [inline src/v0/destinations/marketo_static_list/transform.js:135]: Remove this tag.

Comment (mihir-4116) [reply src/v0/destinations/marketo_static_list/transform.js]: done

**Review by saikumarrs — APPROVED** [review]

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js (hunk 5 above)
Comment (utsabc) [reply src/v0/destinations/marketo_bulk_upload/fileUpload.js]: No customer has to fix it

Comment (saikumarrs) [reply src/v0/destinations/marketo_bulk_upload/fileUpload.js]: @mihir-4116 , let's fix this error type.

**Pull request merged**